executor/union_scan.go

```diff
@@ -286,30 +339,38 @@ func (us *UnionScanExec) rowWithColsInTxn(t table.Table, h int64, cols []*table.
 	if err != nil {
 		return nil, err
 	}
-	v, _, err := tables.DecodeRawRowData(us.ctx, t.Meta(), h, cols, value)
+	colIDs := make(map[int64]int)
+	for i, col := range us.columns {
+		colIDs[col.ID] = i
+	}
+	return decodeRowData(us.ctx, us.table.Meta(), us.columns, colIDs, h, []byte{}, value)
+}
+
+func (us *UnionScanExec) getMemRow(h int64) ([]types.Datum, error) {
+	data, err := us.rowWithColsInTxn(us.table, h)
+	if err != nil {
+		return nil, err
+	}
+	mutableRow := chunk.MutRowFromTypes(us.retTypes())
+	mutableRow.SetDatums(data...)
+	matched, _, err := expression.EvalBool(us.ctx, us.conditions, mutableRow.ToRow())
 	if err != nil {
 		return nil, err
 	}
-	return v, nil
+	if !matched {
+		return nil, nil
+	}
+	return data, nil
 }
 
 func (us *UnionScanExec) buildAndSortAddedRows(t table.Table) error {
```

Comment:
Could we remove it now?

Reply:
How about remove `buildAndSortAddedRows` functions and `DirtyTable`in the next PR? already add a `TODO` here.